clippy_lints/src/methods/result_map_or_else_none.rs

```diff
@@ -0,0 +1,75 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::snippet;
+use clippy_utils::ty::is_type_diagnostic_item;
+use clippy_utils::{is_res_lang_ctor, path_res};
+use rustc_errors::Applicability;
+use rustc_hir as hir;
+use rustc_hir::LangItem::OptionSome;
+use rustc_hir::QPath;
+use rustc_lint::LateContext;
+use rustc_span::symbol::sym;
+
+use super::RESULT_MAP_OR_INTO_OPTION;
+
+// The expression inside a closure may or may not have surrounding braces
+// which causes problems when generating a suggestion.
+fn is_only_none<'a>(cx: &LateContext<'_>, expr: &'a hir::Expr<'_>) -> bool {
+    match expr.kind {
+        hir::ExprKind::Block(block, _) => {
+            match (block.stmts, block.expr) {
+                (&[], Some(inner_expr)) => {
+                    // If block only contains an expression,
+                    // reduce `|x| { x + 1 }` to `|x| x + 1`
+                    is_only_none(cx, inner_expr)
+                },
+                _ => false,
+            }
+        },
+        hir::ExprKind::Path(qpath) => {
+            if let QPath::Resolved(_, path) = qpath
+                && let Some(last) = path.segments.last()
+                && last.ident.name.as_str() == "None"
```

Comment:
I'm a bit surprised that there's no `OptionNone` lang item, while `OptionSome` exists.

Reply:
That will make things much easier, thanks!